spring-kafka/src/main/java/org/springframework/kafka/AsyncKafkaTemplate.java

```diff
@@ -0,0 +1,229 @@
+ *
+ * @param <K> the key type.
+ * @param <V> the outbound data type.
+ * @param <R> the reply data type.
+ *
+ * @author Gary Russell
+ * @since 2.1.3
+ *
+ */
+public class AsyncKafkaTemplate<K, V, R> extends KafkaTemplate<K, V> implements BatchMessageListener<K, R>,
+		InitializingBean, SmartLifecycle {
+
+	private static final long DEFAULT_REPLY_TIMEOUT = 5000L;
+
+	private final GenericMessageListenerContainer<K, R> replyContainer;
+
+	private final ConcurrentMap<String, RequestReplyFuture<K, V, R>> futures = new ConcurrentHashMap<>();
+
+	private TaskScheduler scheduler;
+
+	private boolean running;
+
+	private int phase;
+
+	private boolean autoStartup;
+
+	private long replyTimeout = DEFAULT_REPLY_TIMEOUT;
+
+	public AsyncKafkaTemplate(ProducerFactory<K, V> producerFactory,
+			GenericMessageListenerContainer<K, R> replyContainer) {
+		this(producerFactory, replyContainer, false);
+	}
+
+	public AsyncKafkaTemplate(ProducerFactory<K, V> producerFactory,
+			GenericMessageListenerContainer<K, R> replyContainer, boolean autoFlush) {
+		super(producerFactory, autoFlush);
+		this.replyContainer = replyContainer;
+		this.replyContainer.setupMessageListener(this);
+	}
+
+	protected void setTaskScheduler(TaskScheduler scheduler) {
+		this.scheduler = scheduler;
+	}
+
+	protected void setReplyTimeout(long replyTimeout) {
+		Assert.isTrue(replyTimeout >= 0, "'replyTimeout' must be >= 0");
+		this.replyTimeout = replyTimeout;
+	}
+
+	@Override
+	public void afterPropertiesSet() throws Exception {
+		Assert.notNull(this.scheduler, "A scheduler is required for timeouts");
+	}
+
+	@Override
+	public synchronized void start() {
+		try {
+			afterPropertiesSet();
+		}
+		catch (Exception e) {
+			throw new KafkaException("Failed to initialize", e);
+		}
+		this.replyContainer.start();
+		this.running = true;
+	}
+
+	@Override
+	public synchronized void stop() {
+		this.running = false;
+		this.replyContainer.stop();
+	}
+
+	@Override
+	public synchronized boolean isRunning() {
+		return this.running;
+	}
+
+	@Override
+	public int getPhase() {
+		return this.phase;
+	}
+
+	public void setPhase(int phase) {
+		this.phase = phase;
+	}
+
+	@Override
+	public boolean isAutoStartup() {
+		return this.autoStartup;
+	}
+
+	public void setAutoStartup(boolean autoStartup) {
+		this.autoStartup = autoStartup;
+	}
+
+	@Override
+	public void stop(Runnable callback) {
+		stop();
+		callback.run();
+	}
+
+	public RequestReplyFuture<K, V, R> sendAndReceive(ProducerRecord<K, V> record) {
+		String correlationId = createCorrelationId();
+		record.headers()
+				.add(new RecordHeader(KafkaHeaders.CORRELATION_ID, correlationId.getBytes(StandardCharsets.UTF_8)));
+		if (this.logger.isDebugEnabled()) {
+			this.logger.debug("Sending: " + record + " with correlationId: " + correlationId);
+		}
+		RequestReplyFuture<K, V, R> future = new RequestReplyFuture<>();
+		this.futures.put(correlationId, future);
+		future.setSendFuture(send(record));
```

Comment:
What if we couldn't send here? 
Can't we perform `this.futures.put()` already after this `send()`?
Or do you mean there might be a race condition that reply will come earlier than we perform `this.futures.put()` ?

Thanks

Reply:
Yes (race condition).

Reply:
OK. So, the next schedule must be in the `final` block for the case when we got an exception from the `send(record)`. Or even better do not schedule at all, but just remove in the `catch`. Just because we haven't sent record therefore we won't anticipate any reply at all.

Does it make sense ?